File: forms/provider-transfer-request/full-stack-with-loco-tera-htmx-alpine/tests/engine/transfer_validator_test.rs
use provider_transfer_request_tera_crate::engine::transfer_validator::validate_transfer;
use provider_transfer_request_tera_crate::engine::types::*;
use provider_transfer_request_tera_crate::engine::validation_rules::all_rules;

/// A fully populated, mandatory-complete transfer (acknowledgement not yet
/// started — that's intentional, to keep the acknowledgement rules outside
/// the applicable set in the baseline).
fn fully_completed() -> AssessmentData {
    AssessmentData {
        requesting_provider: ProviderDetails {
            clinician_name: "Dr Alice Hart".to_string(),
            clinician_role: "Registrar".to_string(),
            organisation: "St James's Hospital".to_string(),
            ward: "AMU".to_string(),
            phone: "0113 123 4567".to_string(),
            email: "alice.hart@example.org".to_string(),
            registration_body: "GMC".to_string(),
            registration_number: "GMC 1234567".to_string(),
        },
        receiving_provider: ProviderDetails {
            clinician_name: "Dr Bilal Khan".to_string(),
            clinician_role: "Consultant".to_string(),
            organisation: "Leeds General Infirmary".to_string(),
            ward: "HDU".to_string(),
            phone: "0113 765 4321".to_string(),
            email: String::new(),
            registration_body: String::new(),
            registration_number: String::new(),
        },
        patient_demographics: PatientDemographics {
            first_name: "Jane".to_string(),
            last_name: "Doe".to_string(),
            date_of_birth: "1972-04-11".to_string(),
            sex: "female".to_string(),
            nhs_number: "943 476 5919".to_string(),
            hospital_number: String::new(),
            address_line: "1 High Street".to_string(),
            postcode: "LS1 1AA".to_string(),
            next_of_kin_name: "John Doe".to_string(),
            next_of_kin_phone: "07700 900000".to_string(),
        },
        situation: Situation {
            reason_for_transfer: "Requires HDU monitoring.".to_string(),
            primary_diagnosis: "Acute pancreatitis".to_string(),
            urgency: "urgent".to_string(),
            transfer_type: "inter-hospital".to_string(),
            requested_date_time: "2026-06-01T10:00".to_string(),
        },
        background: Background {
            presenting_complaint: "Severe epigastric pain.".to_string(),
            relevant_history: "Admitted 24h ago.".to_string(),
            past_medical_history: "Cholelithiasis.".to_string(),
            current_medications: "Paracetamol 1g QDS.".to_string(),
            allergies: "NKDA".to_string(),
            recent_investigations: "Amylase 1200; CT pancreas pending.".to_string(),
            infection_status: "No known colonisation.".to_string(),
        },
        assessment: Assessment {
            current_clinical_status: "Comfortable on opioid analgesia.".to_string(),
            conscious_level: "awake".to_string(),
            vital_signs: VitalSigns {
                heart_rate: Some(82.0),
                respiratory_rate: Some(16.0),
                systolic_blood_pressure: Some(118.0),
                diastolic_blood_pressure: Some(76.0),
                temperature_celsius: Some(37.0),
                oxygen_saturation: Some(98.0),
                news_score: Some(1.0),
            },
            clinically_stable: "yes".to_string(),
            stability_notes: String::new(),
        },
        recommendation: Recommendation {
            requested_action: "Accept for HDU monitoring.".to_string(),
            expected_outcomes: "Resolution of pancreatitis.".to_string(),
            ongoing_care_plan: "Hourly observations; IV fluids; analgesia.".to_string(),
            pending_results: "CT pancreas; LFTs at 24h.".to_string(),
        },
        transfer_logistics: TransferLogistics {
            transport_mode: "ambulance".to_string(),
            departure_date_time: "2026-06-01T10:30".to_string(),
            estimated_arrival_date_time: "2026-06-01T11:15".to_string(),
            escort_required: false,
            escort_details: String::new(),
            oxygen_required: false,
            cardiac_monitoring_required: false,
            infectious_precautions: false,
            infectious_precautions_details: String::new(),
            falls_risk: false,
            mental_capacity_concerns: false,
            equipment_required: "Peripheral IV (right ACF).".to_string(),
        },
        signoff_acknowledgement: SignoffAcknowledgement {
            requesting_provider_signature: "Dr Alice Hart".to_string(),
            requesting_provider_signature_date: "2026-06-01".to_string(),
            ..Default::default()
        },
    }
}

#[test]
fn empty_assessment_is_incomplete() {
    let data = AssessmentData::default();
    let result = validate_transfer(&data);
    assert_eq!(result.completeness, "incomplete");
    assert_eq!(result.total_satisfied, 0);
    assert!(result.mandatory_required > 0);
}

#[test]
fn fully_completed_is_complete() {
    let data = fully_completed();
    let result = validate_transfer(&data);
    assert_eq!(result.completeness, "complete");
    assert_eq!(result.total_satisfied, result.total_required);
    assert_eq!(result.mandatory_satisfied, result.mandatory_required);
}

#[test]
fn missing_optional_is_partial() {
    let mut data = fully_completed();
    // Drop a single optional field.
    data.background.past_medical_history = String::new();
    data.recommendation.expected_outcomes = String::new();
    let result = validate_transfer(&data);
    assert_eq!(result.completeness, "partial");
    assert_eq!(result.mandatory_satisfied, result.mandatory_required);
    assert!(result.total_satisfied < result.total_required);
}

#[test]
fn missing_mandatory_is_incomplete() {
    let mut data = fully_completed();
    data.recommendation.requested_action = String::new();
    let result = validate_transfer(&data);
    assert_eq!(result.completeness, "incomplete");
    assert!(result.missing.iter().any(|m| m.id == "REC-01"));
}

#[test]
fn unstable_requires_stability_notes() {
    let mut data = fully_completed();
    data.assessment.clinically_stable = "no".to_string();
    data.assessment.stability_notes = String::new();
    let result = validate_transfer(&data);
    assert!(result.missing.iter().any(|m| m.id == "ASS-04"));
}

#[test]
fn escort_required_requires_escort_details() {
    let mut data = fully_completed();
    data.transfer_logistics.escort_required = true;
    data.transfer_logistics.escort_details = String::new();
    let result = validate_transfer(&data);
    assert!(result.missing.iter().any(|m| m.id == "LOG-04"));
}

#[test]
fn acknowledgement_gating_only_fires_once_touched() {
    let baseline = fully_completed();
    let baseline_result = validate_transfer(&baseline);
    assert!(!baseline_result.missing.iter().any(|m| m.id == "SGN-03"));

    let mut started = baseline.clone();
    started.signoff_acknowledgement.receiving_provider_name = "Dr Bilal Khan".to_string();
    let started_result = validate_transfer(&started);
    // Once the name is filled in, signature / signature date / checkbox fire.
    assert!(started_result.missing.iter().any(|m| m.id == "SGN-04"));
    assert!(started_result.missing.iter().any(|m| m.id == "SGN-05"));
    assert!(started_result.missing.iter().any(|m| m.id == "SGN-06"));
}

#[test]
fn rule_ids_are_unique() {
    let rules = all_rules();
    let mut ids: Vec<&str> = rules.iter().map(|r| r.id).collect();
    ids.sort();
    let len_before = ids.len();
    ids.dedup();
    assert_eq!(ids.len(), len_before, "rule IDs must be unique");
}

File: forms/provider-transfer-request/front-end-with-svelte/src/lib/engine/transfer-validator.test.ts
import { describe, it, expect } from 'vitest';
import { validateTransfer, gradeTransfer } from './transfer-validator';
import { detectFlaggedIssues } from './flagged-issues';
import { validationRules, acknowledgementStarted } from './validation-rules';
import { createDefaultAssessment } from './defaults';
import type { AssessmentData } from './types';

/** A fully-completed transfer with all mandatory fields supplied. */
function completeTransfer(): AssessmentData {
	const d = createDefaultAssessment();
	d.requestingProvider = {
		...d.requestingProvider,
		clinicianName: 'Dr A Requester',
		clinicianRole: 'Registrar',
		organisation: 'St Mary General',
		ward: 'AMU',
		phone: '01000 000001',
		email: 'a.req@example.org'
	};
	d.receivingProvider = {
		...d.receivingProvider,
		clinicianName: 'Dr B Receiver',
		clinicianRole: 'Consultant',
		organisation: 'County Cardiac Centre',
		ward: 'CCU',
		phone: '01000 000002'
	};
	d.patientDemographics = {
		...d.patientDemographics,
		firstName: 'Pat',
		lastName: 'Patient',
		dateOfBirth: '1960-01-01',
		sex: 'female',
		nhsNumber: '123 456 7890',
		nextOfKinName: 'Kin Person'
	};
	d.situation = {
		...d.situation,
		reasonForTransfer: 'Specialist cardiac input required.',
		primaryDiagnosis: 'NSTEMI',
		urgency: 'routine',
		transferType: 'inter-hospital',
		requestedDateTime: '2026-07-01T09:00'
	};
	d.background = {
		...d.background,
		presentingComplaint: 'Chest pain',
		relevantHistory: 'Admitted overnight, troponin positive.',
		pastMedicalHistory: 'Hypertension',
		currentMedications: 'Aspirin 75mg OD',
		allergies: 'NKDA',
		recentInvestigations: 'ECG, troponin',
		infectionStatus: 'None known'
	};
	d.assessment = {
		...d.assessment,
		currentClinicalStatus: 'Stable, pain-free.',
		consciousLevel: 'awake',
		clinicallyStable: 'yes',
		vitalSigns: {
			...d.assessment.vitalSigns,
			heartRate: 78,
			respiratoryRate: 16,
			systolicBloodPressure: 128,
			oxygenSaturation: 98
		}
	};
	d.recommendation = {
		...d.recommendation,
		requestedAction: 'Accept for angiography.',
		expectedOutcomes: 'Revascularisation.',
		ongoingCarePlan: 'Continue dual antiplatelet therapy.',
		pendingResults: 'None'
	};
	d.transferLogistics = {
		...d.transferLogistics,
		transportMode: 'ambulance',
		departureDateTime: '2026-07-01T10:00',
		estimatedArrivalDateTime: '2026-07-01T10:45'
	};
	d.signoffAcknowledgement = {
		...d.signoffAcknowledgement,
		requestingProviderSignature: 'A Requester',
		requestingProviderSignatureDate: '2026-07-01',
		receivingProviderName: 'Dr B Receiver',
		receivingProviderSignature: 'B Receiver',
		receivingProviderSignatureDate: '2026-07-01',
		acknowledgementReceived: true
	};
	return d;
}

describe('Provider Transfer completeness validator', () => {
	it('reports incomplete for a blank transfer (mandatory fields missing)', () => {
		const result = validateTransfer(createDefaultAssessment());
		expect(result.completeness).toBe('incomplete');
		expect(result.mandatorySatisfied).toBeLessThan(result.mandatoryRequired);
		expect(result.missing.length).toBeGreaterThan(0);
	});

	it('reports complete when every applicable field is supplied', () => {
		const result = validateTransfer(completeTransfer());
		expect(result.completeness).toBe('complete');
		expect(result.totalSatisfied).toBe(result.totalRequired);
		expect(result.missing).toHaveLength(0);
	});

	it('reports partial when only optional fields are outstanding', () => {
		const d = completeTransfer();
		d.requestingProvider.email = '';
		d.requestingProvider.ward = '';
		const result = validateTransfer(d);
		expect(result.completeness).toBe('partial');
		expect(result.mandatorySatisfied).toBe(result.mandatoryRequired);
		expect(result.missing.every((m) => !m.mandatory)).toBe(true);
	});

	it('requires stability notes only when not clinically stable', () => {
		const d = completeTransfer();
		d.assessment.clinicallyStable = 'no';
		d.assessment.stabilityNotes = '';
		const result = validateTransfer(d);
		expect(result.completeness).toBe('incomplete');
		expect(result.missing.some((m) => m.id === 'ASS-04')).toBe(true);
	});

	it('gates receiving-provider rules until acknowledgement is started', () => {
		const blank = createDefaultAssessment();
		expect(acknowledgementStarted(blank)).toBe(false);
		const ackRules = validationRules.filter((r) => r.id.startsWith('SGN-0') && r.id !== 'SGN-01' && r.id !== 'SGN-02');
		expect(ackRules.every((r) => !r.applies(blank))).toBe(true);

		const started = completeTransfer();
		started.signoffAcknowledgement.receivingProviderName = 'Dr B Receiver';
		expect(acknowledgementStarted(started)).toBe(true);
		expect(ackRules.every((r) => r.applies(started))).toBe(true);
	});

	it('has unique rule ids', () => {
		const ids = validationRules.map((r) => r.id);
		expect(new Set(ids).size).toBe(ids.length);
	});
});

describe('Provider Transfer flagged issues', () => {
	it('raises no flags for a stable, routine transfer', () => {
		expect(detectFlaggedIssues(completeTransfer())).toHaveLength(0);
	});

	it('flags emergent urgency as urgent priority', () => {
		const d = completeTransfer();
		d.situation.urgency = 'emergent';
		const flags = detectFlaggedIssues(d);
		expect(flags.some((f) => f.id === 'FLAG-URG-EMERG' && f.priority === 'urgent')).toBe(true);
	});

	it('flags a clinically unstable patient', () => {
		const d = completeTransfer();
		d.assessment.clinicallyStable = 'no';
		d.assessment.stabilityNotes = 'Hypotensive.';
		const flags = detectFlaggedIssues(d);
		expect(flags.some((f) => f.id === 'FLAG-STAB-NO')).toBe(true);
	});

	it('flags critically low oxygen saturation', () => {
		const d = completeTransfer();
		d.assessment.vitalSigns.oxygenSaturation = 85;
		const flags = detectFlaggedIssues(d);
		expect(flags.some((f) => f.id === 'FLAG-VIT-SPO2')).toBe(true);
	});

	it('sorts flags by priority (urgent first)', () => {
		const d = completeTransfer();
		d.situation.urgency = 'emergent';
		d.transferLogistics.fallsRisk = true;
		const flags = detectFlaggedIssues(d);
		const order = { urgent: 0, high: 1, medium: 2, low: 3 };
		const priorities = flags.map((f) => order[f.priority]);
		expect(priorities).toEqual([...priorities].sort((a, b) => a - b));
	});
});

describe('gradeTransfer', () => {
	it('bundles validation, flags, and a timestamp', () => {
		const g = gradeTransfer(completeTransfer());
		expect(g.validation.completeness).toBe('complete');
		expect(Array.isArray(g.flags)).toBe(true);
		expect(typeof g.timestamp).toBe('string');
	});
});
